Differential gene expression in mice (germ free or normal, different ages, 2 genders)
=====================================================================================

# Biological context

## Questions

- What are the differentially expressed genes of the microglya cells between germ free (GF) and conventional (SPF) mice at different ages?
- Is there a difference between the ages?
- Is there a difference between the genders?
- Which genes? Which pathways?

## Hypotheses

- Differential gene expression between SPF and GF already known for 8 weeks mices ([article](http://www.nature.com/neuro/journal/v18/n7/abs/nn.4030.html))
- Expected gender bias in 104w SPF
- Differential gene expression between SPF and GF already known for embryonic and early mices ([article](https://www.sciencedirect.com/science/article/pii/S0092867417314320))

# Data

Microglia cells have been extracted from 64 mices given:
- 2 microbiota states: Conventional (SPF) and Germ free (GF)
- 3 ages: Young (8 weeks), Middle-aged (52 weeks), Old (104 weeks)
- 2 sexes: Female and Male 

Microbiote | Age | Sex | Number of samples
--- | --- | --- | --
GF | Middle-aged | Female | 6
GF | Middle-aged | Male | 4
GF | Old | Female | 3
GF | Old | Male | 5
GF | Young | Female | 5
GF | Young | Male | 4
SPF | Middle-aged | Female | 6
SPF | Middle-aged | Male | 5
SPF | Old | Female | 3
SPF | Old | Male | 14
SPF | Young | Female | 5
SPF | Young | Male | 4

Total RNA was extracted from FACS sorted CD11b+CD45lowLin- microglia cells using The ARCTURUS® PicoPure® RNA Isolation Kit (ThermoFisher) according to manufacturer’s protocol. The SMARTer Ultra Low Input RNA Kit for Sequencing v4 (Clontech Laboratories, Inc., Mountain View, CA, USA) was used to generate first strand cDNA from 500 to 750 pg total-RNA. Double stranded cDNA was amplified by LD PCR (11 cycles) and purified via magnetic bead clean-up.

Library preparation was carried out as described in the Illumina Nextera XT Sample Preparation Guide (Illumina, Inc., San Diego, CA, USA). 150 pg of input cDNA were tagmented (tagged and fragmented) by the Nextera XT transposome. The products were purified and amplified via a limited-cycle PCR program to generate multiplexed sequencing libraries. For the PCR step 1:5 dilutions of index 1 (i7) and index 2 (i5) primers were used. The libraries were quantified using the KAPA SYBR FAST ABI Prism Library Quantification Kit (Kapa Biosystems, Inc., Woburn, MA, USA). Equimolar amounts of each library were pooled, and the pools were used for cluster generation on the cBot with the Illumina TruSeq SR Cluster Kit v3.

The sequencing run was performed on a HiSeq 1000 instrument using the indexed, 50 cycles single-read (SR) protocol and the TruSeq SBS v3 Reagents according to the Illumina HiSeq 1000 System User Guide. Image analysis and base calling were converted into FASTQ files with the CASAVA1.8.2 software. Library preparation and RNAseq were performed at the Genomics Core Facility "KFB - Center of Excellence for Fluorescent Bioanalytics" (University of Regensburg, Regensburg, Germany).

# Analyses

## Extraction of read counts of the each genes for each samples

extraction of number of reads mapped on each annotated genes for each samples
[details](gene_count_extraction)

## Preparation of the differential expression analysis

Count data was first cleaned: 3 samples with low mapping rates (< 75%) or low number and rates of reads assigned to genes (< 55%), genes not find in any samples were removed, the genes names were checked and the metadata extracted and formatted. The details can be found [here](prepare_data). 

The **differential expression analysis** is performed with DESeq2 v.1.14.1 {% cite Love2014 %}. The design model integrates the different factors (sex, microbiota and age) and their interaction: `Sex + Microbiota + Age + Sex:Age + Sex:Microbiota + Microbiota:Age`. The addition of each factor and interactions were tested and the number of genes with a significant adjusted p-values for the LRT (Likelihood Ratio Tests) were reported to show the percentage of variables' effect on transcriptomic profile. The details of this report as well as exploratory analysis and visualization of variance stabilizing transformation of the DESeq2 model can be found [here](dge_analysis).

Normalized counts generated by DESeq2 were checked for artefacts due to FACS sorting or contamination from other cell types that may have escaped the sorting gating. 3 samples and 30 markers genes of non microglia cell types were then removed.

**Clustering of the samples** based on the normalized counts was performed using hierarchical Ward clustering ({% cite Murtagh2014 %}), implemented in the R (v.3.4.3) `stats` package to generate the dendogram ([Figure S1A](pre-visualization#With-all-genes)). [**Principal Components Analyses**](pre-visualization#With-all-genes#PCA-on-the-normalized-counts) were performed using the R (v.3.4.3) `stats` package: 1. on the full normalized counts and with several color codes to inspect the data; 2. to compare the 2 microbiote states for the different combinations of age and sex.

[**Weighted gene co-expression network analysis**](pre-visualization#Gene-co-expression-analysis) (WGCNA, {% cite Zhang2005 %})  was performed on the normalized expression data using the R package WGCNA v1.63 ({% cite Langfelder2008 %}). For computational efficiency, genes were filtered to keep only genes that have a at least 10 counts in more than 90% of the samples (10277 genes, 9417 removed). To obtain a signed hybrid network fulfilling the scale free topology, the soft-thresholding power parameter was set to 6. Co-expression modules were defined using a minimum module size of 65 genes and by merging modules with a module eigengene dissimilarity below 0.35 resulting in 9 modules having sizes between 117 and 1627 genes.

A module-trait correlation analysis was performed between the module eigengene (ME) and the different trait (combination of microbiote, age and sex): correlation of Pearson between each pair of variables and Student asymptotic p-values for the correlations, computed using WGCNA package.


The modules were characterized by GO enrichment with a focus on biological processes using the R package clusterProfiler v3.0.5 (Yu et al., 2012). The co-expression network was generated based on the Topological Overlap Matrix requiring an edge weight of more than 0.3 to at least one other gene and excluding genes belonging to the gray module that contains those genes that do not fit in any other module. Network visualization was carried out in Cytoscape v3.4.0 applying the edge-weighted spring-embedded layout and hiding smaller clusters with less than 5 genes. In the base network nodes were colored according to module membership while in all other networks DEGs (Fold- Change > 1.5 or < 1.5 and CPM > 25) were highlighted by red (positive FC) or blue (negative FC) fill color. Transcription factors were marked by a triangular node shape.


Before any differential analysis, we [pre-visualized of the normalized count data ](pre-visualization) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/modules/)

tibble_1.4.1               dplyr_0.7.4               
 [3] sinaplot_1.1.0             sva_3.22.0                
 [5] genefilter_1.56.0          mgcv_1.8-22               
 [7] nlme_3.1-131               gridExtra_2.3             
 [9] ggfortify_0.4.5            RamiGO_1.20.0             
[11] gsubfn_0.7                 proto_1.0.0               
[13] plotly_4.7.1               reshape_0.8.7             
[15] limma_3.30.13              xlsx_0.5.7                
[17] xlsxjars_0.6.1             rJava_0.9-9               
[19] pathview_1.14.0            org.Hs.eg.db_3.4.0        
[21] htmlwidgets_0.9            threejs_0.3.1             
[23] clusterProfiler_3.2.14     DOSE_3.0.10               
[25] igraph_1.2.1               WGCNA_1.63                
[27] fastcluster_1.1.24         dynamicTreeCut_1.63-1     
[29] moe430a.db_3.2.3           annotate_1.52.1           
[31] XML_3.98-1.9               GO.db_3.4.0               
[33] reshape2_1.4.3             GOSemSim_2.0.4            
[35] dendextend_1.8.0           goseq_1.26.0              
[37] geneLenDataBase_1.10.0     BiasedUrn_1.07            
[39] rentrez_1.2.1              org.Mm.eg.db_3.4.0        
[41] AnnotationDbi_1.36.2       viridis_0.5.0             
[43] viridisLite_0.3.0          RColorBrewer_1.1-2        
[45] plyr_1.8.4                 ggplot2_2.2.1             
[47] UpSetR_1.3.3               gplots_3.0.1              
[49] pheatmap_1.0.8             DESeq2_1.14.1             
[51] SummarizedExperiment_1.4.0 Biobase_2.34.0            
[53] GenomicRanges_1.26.4       GenomeInfoDb_1.10.3       
[55] IRanges_2.8.2              S4Vectors_0.12.2          
[57] BiocGenerics_0.20.0   

## Differential expression analysis

We then analyzed the data given different angles

### Effect of the type on the expressed genes

Analyses:
1. [Comparison between the types (after controlling for age and gender)](type-effect-general) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/type-effect/type)
2. [Comparison between the types for the genders (after controlling for age)](type-effect-gender) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/type-effect/type_gender)
3. [Comparison between the types for the ages (after controlling for gender)](type-effect-age) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/type-effect/type_age)
4. [Comparison between the types for the genders and ages](type-effect-age-gender) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/type-effect/type_gender_age)

### Effect of the gender on the expressed genes

Analyses:
1. [Comparison between the genders (after controlling for type and age)](gender-effect-general) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/gender-effect/gender)
2. [Comparison between the genders for the types (after controlling for age)](gender-effect-type) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/gender-effect/gender_type)
3. [Comparison between the genders for the ages (after controlling for type)](gender-effect-age) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/gender-effect/gender_age)
4. [Comparison between the genders for the types and ages](gender-effect-type-age) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/gender-effect/gender_type_age)

### Effect of the age on the expressed genes

Analyses:
1. [Comparison between the ages (after controlling for type and gender)](age-effect-general) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/age-effect/age)
2. [Comparison between the ages for the types (after controlling for gender)](age-effect-type) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/age-effect/age_type)
3. [Comparison between the ages for the genders (after controlling for type)](age-effect-gender) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/age-effect/age_gender)
4. [Comparison between the ages for the genders and types](age-effect-type-gender) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/age-effect/age_type_gender)
5. [Comparison with results of Grabert et al, 2016](grabert_comparison)

## Genes confirmed as a core signature for homeostatic microglia

[Genes confirmed as a core signature for homeostatic microglia](homeostatic_core_signature_genes) - [data](https://github.com/bebatut/neuromac_GF_mices/tree/master/results/dge/homeostatic_core_signature_genes/)

## Comparison of GF and SPF 8w Female to GF 8w Female with added microbiota (young or old)

[Comparison of GF and SPF 8w Female to GF 8w Female with added microbiota (young or old)](added_microbiota)